File: scripts/backend-python/paytm_service.py
"""Paytm server-side flow: initiateTransaction, order status, callback verify.

Mirrors backend-node/paytmService.js. Uses Paytm's official `paytmchecksum` package.
"""
import json
import secrets
from decimal import Decimal, ROUND_HALF_UP
from typing import Any, Optional

import requests
from paytmchecksum import PaytmChecksum

from paytm_config import get_paytm_config


class PaytmError(Exception):
    def __init__(self, code: str, message: str, *, http_status: int = 500,
                 order_id: Optional[str] = None, paytm: Optional[dict] = None):
        super().__init__(message)
        self.code = code
        self.http_status = http_status
        self.order_id = order_id
        self.paytm = paytm


_TWO_PLACES = Decimal("0.01")
_FALLBACK_AMT = Decimal("1.00")


def _normalize_amount(amount: Any) -> str:
    """Two-decimal currency normalization using Decimal (avoids binary-float drift).

    Paytm's `txnAmount.value` is a STRING with two decimals, so we serialize
    via Decimal's plain string form. Float arithmetic is never used.
    """
    raw = ("" if amount is None else str(amount)).strip()
    try:
        d = Decimal(raw) if raw else _FALLBACK_AMT
    except Exception:
        d = _FALLBACK_AMT
    if d <= 0:
        d = _FALLBACK_AMT
    return str(d.quantize(_TWO_PLACES, rounding=ROUND_HALF_UP))


def _require_credentials() -> dict:
    cfg = get_paytm_config()
    if not cfg["mid"]:
        raise PaytmError("MISSING_MID", "Missing PAYTM_MID")
    if not cfg["merchant_key"]:
        raise PaytmError("MISSING_MERCHANT_KEY", "Missing PAYTM_MERCHANT_KEY")
    return cfg


def initiate_transaction(
    amount: Any,
    cust_id: Optional[str],
    server_base_url: str,
    mobile: Optional[str] = None,
    email: Optional[str] = None,
    order_id: Optional[str] = None,
) -> dict:
    cfg = _require_credentials()
    # Accept a merchant-supplied orderId for reconciliation; fall back to a random one.
    order_id = (order_id or "").strip() or ("ORD_" + secrets.token_hex(10).upper())
    callback_url = cfg["callback_url"] or f"{server_base_url.rstrip('/')}/paytm/callback"

    user_info = {"custId": (cust_id or "").strip() or "CUST_DEMO"}
    # mobile + email are strongly recommended — pre-fill the consent screen and
    # drive OTP / notifications. Real merchants should always pass these through.
    if mobile and mobile.strip():
        user_info["mobile"] = mobile.strip()
    if email and email.strip():
        user_info["email"] = email.strip()

    body = {
        "requestType": "Payment",
        "mid": cfg["mid"],
        "websiteName": cfg["website_name"],
        "orderId": order_id,
        "callbackUrl": callback_url,
        "txnAmount": {"value": _normalize_amount(amount), "currency": "INR"},
        "userInfo": user_info,
    }
    signature = PaytmChecksum.generateSignature(json.dumps(body), cfg["merchant_key"])
    payload = {"body": body, "head": {"signature": signature}}

    url = f"{cfg['initiate_transaction_url']}?mid={cfg['mid']}&orderId={order_id}"
    r = requests.post(url, json=payload, headers={"Content-Type": "application/json"}, timeout=15)
    text = r.text
    if not r.ok:
        raise _upstream("INITIATE_HTTP_ERROR", f"initiateTransaction failed (HTTP {r.status_code})", order_id, text)

    data = json.loads(text)
    info = (data.get("body") or {}).get("resultInfo") or {}
    if info.get("resultStatus") != "S":
        raise _upstream("INITIATE_FAILED", info.get("resultMsg") or "initiateTransaction failed", order_id, text)
    txn_token = (data.get("body") or {}).get("txnToken")
    if not txn_token:
        raise _upstream("MISSING_TXN_TOKEN", "Missing txnToken in Paytm response", order_id, text)

    return {
        "orderId": order_id,
        "txnToken": txn_token,
        "amount": body["txnAmount"]["value"],
        "mid": cfg["mid"],
        "tokenType": "TXN_TOKEN",
    }


def fetch_order_status(order_id: str) -> str:
    cfg = _require_credentials()
    body = {"mid": cfg["mid"], "orderId": order_id}
    signature = PaytmChecksum.generateSignature(json.dumps(body), cfg["merchant_key"])
    payload = {"body": body, "head": {"signature": signature}}
    r = requests.post(cfg["order_status_url"], json=payload,
                      headers={"Content-Type": "application/json"}, timeout=15)
    if not r.ok:
        raise PaytmError("STATUS_HTTP_ERROR", f"order status HTTP {r.status_code} — {r.text}",
                         http_status=502, order_id=order_id)
    return r.text


def verify_callback_checksum(params: dict) -> bool:
    cfg = get_paytm_config()
    signature = params.get("CHECKSUMHASH")
    if not signature or not cfg["merchant_key"]:
        return False
    to_verify = {k: v for k, v in params.items() if k != "CHECKSUMHASH"}
    try:
        return PaytmChecksum.verifySignature(to_verify, cfg["merchant_key"], signature)
    except Exception:
        return False


def _upstream(code: str, message: str, order_id: str, raw_body: str) -> PaytmError:
    paytm: dict = {}
    try:
        info = ((json.loads(raw_body) or {}).get("body") or {}).get("resultInfo") or {}
        for k in ("resultStatus", "resultCode", "resultMsg"):
            if info.get(k):
                paytm[k] = info[k]
    except Exception:
        pass
    return PaytmError(code, message, http_status=502, order_id=order_id, paytm=paytm or None)

File: scripts/backend-node/paytmService.js
import crypto from "node:crypto";
import PaytmChecksum from "paytmchecksum";
import { getPaytmConfig } from "./paytmConfig.js";

function normalizeAmount(amount) {
  // Two-decimal currency normalization. We round via integer paise (×100,
  // round-half-up via Number.EPSILON, ÷100) to avoid binary-float drift, then
  // emit as a string with exactly two decimals - Paytm's `txnAmount.value`
  // contract.
  const raw = (amount ?? "").toString().trim();
  if (!raw) return "1.00";
  const n = Number(raw);
  if (!Number.isFinite(n) || n <= 0) return "1.00";
  const rounded = Math.round((n + Number.EPSILON) * 100) / 100;
  return rounded.toFixed(2);
}

export function requireCredentials() {
  const cfg = getPaytmConfig();
  if (!cfg.mid) {
    const err = new Error("Missing PAYTM_MID");
    err.code = "MISSING_MID";
    throw err;
  }
  if (!cfg.merchantKey) {
    const err = new Error("Missing PAYTM_MERCHANT_KEY");
    err.code = "MISSING_MERCHANT_KEY";
    throw err;
  }
  return cfg;
}

export async function initiateTransaction({ amount, custId, mobile, email, orderId: callerOrderId, serverBaseUrl }) {
  const cfg = requireCredentials();
  // Accept a merchant-supplied orderId for reconciliation against your own orders table.
  // Falls back to a random 20-hex-char id when not supplied.
  const orderId = callerOrderId?.trim() || `ORD_${crypto.randomBytes(10).toString("hex").toUpperCase()}`;
  const effectiveBase = (serverBaseUrl ?? "").toString().trim() || cfg.callbackBase;
  const callbackUrl = cfg.callbackUrl?.trim() || `${effectiveBase}/paytm/callback`;

  const body = {
    requestType: "Payment",
    mid: cfg.mid,
    websiteName: cfg.websiteName,
    orderId,
    callbackUrl: callbackUrl,
    txnAmount: { value: normalizeAmount(amount), currency: "INR" },
    userInfo: {
      custId: custId?.trim() ? custId.trim() : "CUST_DEMO",
      // mobile + email are strongly recommended - pre-fill the consent screen and
      // drive OTP / notifications. Real merchants should always pass these through.
      ...(mobile?.trim() ? { mobile: mobile.trim() } : {}),
      ...(email?.trim() ? { email: email.trim() } : {}),
    },
  };

  const signature = await PaytmChecksum.generateSignature(JSON.stringify(body), cfg.merchantKey);
  const payload = { body, head: { signature } };

  const url = new URL(cfg.initiateTransactionUrl);
  url.searchParams.set("mid", cfg.mid);
  url.searchParams.set("orderId", orderId);

  const r = await fetch(url.toString(), {
    method: "POST",
    headers: { "Content-Type": "application/json" },
    body: JSON.stringify(payload),
  });
  const text = await r.text();
  if (!r.ok) throw asPaytmUpstreamError("INITIATE_HTTP_ERROR", `initiateTransaction failed (HTTP ${r.status})`, orderId, text);

  const json = JSON.parse(text);
  const status = json?.body?.resultInfo?.resultStatus;
  if (status !== "S") {
    const msg = json?.body?.resultInfo?.resultMsg || "initiateTransaction failed";
    throw asPaytmUpstreamError("INITIATE_FAILED", msg, orderId, text);
  }

  const txnToken = json?.body?.txnToken;
  if (!txnToken) throw asPaytmUpstreamError("MISSING_TXN_TOKEN", "Missing txnToken in Paytm response", orderId, text);

  return { orderId, txnToken, amount: body.txnAmount.value, mid: cfg.mid, tokenType: "TXN_TOKEN" };
}

export async function fetchOrderStatus({ orderId }) {
  const cfg = requireCredentials();
  const body = { mid: cfg.mid, orderId };
  const signature = await PaytmChecksum.generateSignature(JSON.stringify(body), cfg.merchantKey);
  const payload = { body, head: { signature } };

  const r = await fetch(cfg.orderStatusUrl, {
    method: "POST",
    headers: { "Content-Type": "application/json" },
    body: JSON.stringify(payload),
  });
  const text = await r.text();
  if (!r.ok) throw new Error(`order status HTTP ${r.status} - ${text}`);
  return text;
}

export function verifyCallbackChecksum(params) {
  const cfg = getPaytmConfig();
  const signature = params?.CHECKSUMHASH;
  if (!signature || !cfg.merchantKey) return false;

  const toVerify = { ...params };
  delete toVerify.CHECKSUMHASH;
  try {
    return PaytmChecksum.verifySignature(toVerify, cfg.merchantKey, signature);
  } catch {
    // Library can throw on malformed/empty input - treat as failed verification.
    return false;
  }
}

function asPaytmUpstreamError(code, message, orderId, rawBody) {
  const e = new Error(message);
  e.code = code;
  e.orderId = orderId;
  e.httpStatus = 502;
  e.paytm = {};
  try {
    const json = JSON.parse(rawBody);
    const info = json?.body?.resultInfo ?? {};
    if (info.resultStatus) e.paytm.resultStatus = info.resultStatus;
    if (info.resultCode) e.paytm.resultCode = info.resultCode;
    if (info.resultMsg) e.paytm.resultMsg = info.resultMsg;
  } catch {
    // ignore
  }
  if (Object.keys(e.paytm).length === 0) delete e.paytm;
  return e;
}
